reasoning/src/commonMain/kotlin/com/xebia/functional/xef/reasoning/tools/ReActAgent.kt

```diff
@@ -171,8 +171,16 @@ class ReActAgent(
       is AgentAction -> {
         logger.info { "🤔 ${plan.thought}" }
         logger.info { "🛠 ${plan.tool}[${plan.toolInput}]" }
+
+        val tool = tools.find { it.name.equals(plan.tool, ignoreCase = true) }
+
         val observation: String? =
-          tools.find { it.name.equals(plan.tool, ignoreCase = true) }?.invoke(plan.toolInput)
+          when (tool) {
+            is ToolWikipedia -> tool.invoke(plan.toolInput, null, null)
```

Comment:
I don't think we should alter the hierarchy of tools with `ToolMain` and introduce a hard check here for this tool.
Instead, it would be better suited as three different tools, one for each capability. Since most use cases are natural language input, the helpful tool is only the search tool. The other two are secondary in case the search tool returns relevant results containing page ids or entity names.
Most use cases won't include users explicitly typing the page id or name but instead, some prompt that will derive a search input. Search is the primary use case, and the others are secondary.

With this in mind, I think a design that does not require hardcoding would be composed of 3 tools:
- Search (primary)
- Page Id (secondary) States in its description that it can only be used with valid Wikipedia page ids returned by the primary search tool
- Page Name (secondary) States in its description that it can only be used with valid Wikipedia page names returned by the primary search tool.

In the end, the ReAct agent should be able to use the secondary tools only if it has first performed a Wikipedia search with the primary tool.

Reply:
Done